### ISSUE TITLE
memory/cpu constraints are no replacement for instanceCap

### ISSUE THREAD

**Issue by fxnn** [jenkinsci/docker-plugin#768]
# versions

docker-plugin 1.1.9
jenkins 2.204.1
docker engine 19.03.5

# problem description

docker-plugin [recommends](https://github.com/jenkinsci/docker-plugin/pull/711/files#diff-9856652c1e5358bdfa1d439ef8e33d2fR6) not using the `instanceCap` parameter, in favor of configuring `memoryLimit` (and the like):

    you should configure template with memory/cpu constraints, so docker infrastructure can manage resource consumption

Due to this notice, I tried hard to enforce memory limits, only to find out that a `memoryLimit` will not stop Jenkins/Docker from starting containers, it will only force the container to use no more memory, thus leading to swapping and finally OOM killing. 

In the end, Jenkins starts so many containers that Jenkins gets OOM killed. Even if it were the builds that were killed, this is no way one can create a stable CI environment.

(I also tried [cgrouping](https://unix.stackexchange.com/questions/537645/how-to-limit-docker-total-resources) all Docker containers, which basically leads to the same thing.)

Therefore I'd propose to remove the misleading deprecation notice from the documentation.

**Comment by pjdarton** [comment]
I'm all in favour of improvements to the help text...
What should it say instead?

**Comment by pjdarton** [comment]
PS. I believe there's more than one set of opinions on this issue.
I'd recommend that you trace back on the history of the existing helptext to see what (if any) rationale was given when changing it to include this "don't do X, do Y instead" advice.
There's usually a good reason behind things and it's bad form to reverse a decision without first understanding it; during my programming career, I've lost count of the number of times bugs have been re-introduced because someone (later) removed a fix because they didn't understand why it was there 😉

FYI The help text is in files called "help-<fieldName>.html" in the src/main/resources tree. e.g. https://github.com/jenkinsci/docker-plugin/tree/docker-plugin-1.1.9/src/main/resources/com/nirima/jenkins/plugins/docker/DockerTemplateBase and https://github.com/jenkinsci/docker-plugin/tree/docker-plugin-1.1.9/src/main/resources/com/nirima/jenkins/plugins/docker/DockerTemplate

Personally, I'm all in favour of instanceCaps and that's what I use where I work _but_ (as I understand it) another option is to use memory/cpu constraints and that'll then "let the docker daemon figure it all out".
It may be (I don't know - I'm just guessing) that the latter could work better with (large) docker swarms, if the swarm performs load balancing based upon memory/cpu constraints.

If you can come up with better wording (or, ideally, a pull request containing that better wording) then I'd be happy to include that in the next release.

**Comment by fxnn** [comment]
I totally agree with you. I already linked the [PR that introduced the line of text](https://github.com/jenkinsci/docker-plugin/pull/711/files#diff-9856652c1e5358bdfa1d439ef8e33d2fR6) in my original issue statement, but unfortunately, I couldn't find and hint there. And it's actually the last commit on the file.

My hope was, that someone here could tell me, why the question isn't wrong and what I was missing. If there's no one, I'd file a PR removing that line.

**Comment by fxnn** [comment]
Oh, regarding the larger docker swarms. To my knowledge, `docker-plugin` can only communicate with "[Standalone Swarm|https://docs.docker.com/swarm/overview/]"; not with "[Swarm Mode|https://docs.docker.com/engine/swarm/]" (as the latter exposes a different API than a regular docker daemon, while standalone swarm supports the same API).

The standalone swarm actually supports [scheduling strategies](https://docs.docker.com/swarm/scheduler/strategy/), and with `binpacking` they _can_ actually consider memory limits and also deny scheduling the container when the host is full.

Unfortunately, AFAIK, standalone swarm is a legacy product and won't be continued, so it doesn't make sense to decide for standalone swarm right now.

**Comment by pjdarton** [comment]
FYI the PR you linked was *me* reformatting/enhancing the help _but_ I didn't originally write that line of text.  So, sure, I touched that line last, but I didn't write it.
Prior to my PR, it looked like [this](https://github.com/pjdarton/docker-plugin/blame/3c26f25d264f6da13a9823662f2237e25f721ba8/src/main/resources/com/nirima/jenkins/plugins/docker/DockerTemplate/help-instanceCapStr.html) which the first version of that file to have that deprecation comment.
The deprecation comment was added in commit 29773ba99604da4e77e3df4b1d42b9bd04d5cab2 ... but unfortunately there's no mention of any reason for that particular change - it looks like that change was bundled in with other (more "structural") changes to the UI.
That was back in 2017, roughly when Nick deLoof took over maintenance of this plugin, bringing it back to life (but it was a bit of a bumpy ride at the time) ... but Nick (the author of that change) has since stepped back from this plugin, so isn't here to offer any insight.
TL;DR: it's unclear exactly what the reasoning was when that was written and it's possible it might have not been very well reasoned (especially with the wisdom of hindsight!).

I suspect you're right w.r.t. swarms - I've never had to get involved in that side of things so I don't know.

So... moving forwards, I think we need to add some extra text to the help text.
Something like "If you are using a swarm that load-balances according to container resource usage then you should set resource limits instead of instance-count limits".
i.e. we shouldn't say "Don't use this it's deprecated", we should say "In <situation> you should use <alternative> instead".
I would welcome a PR making such changes...

**Comment by fxnn** [comment]
Oops, my fault; thank you also for the background information. Your idea about the help text sounds good so far, I'll hopefully find some time to file a PR.

**Comment by pjdarton** [comment]
Thanks for the PR; It'll be included in the next release.
If you can't wait that long, you'll find the latest build here: https://ci.jenkins.io/job/Plugins/job/docker-plugin/job/master/
